Resolves #33 - An authenticated user can update an answer of a given question

adevr opened the pull request.

adevr commented:
It's ready for code review! Please let me know if it needs any adjustment! 👀

tiagof commented on domains/Discussions/Database/Migrations/2020_10_12_104113_create_questions_table.php:
```diff
@@ -11,7 +11,7 @@ public function up(): void
         Schema::create('questions', function (Blueprint $table) {
             $table->id();
-            $table->foreignIdFor(User::class, 'author_id')->constrained();
+            $table->foreignIdFor(User::class, 'author_id')->constrained("users");
```
Well spotted!

josepostiga replied:
Use the single-quote `'`

tiagof commented on domains/Discussions/Controllers/AnswersUpdateController.php:
```diff
@@ -0,0 +1,42 @@
+class AnswersUpdateController extends Controller
+{
+    private Answer $answer;
+    private Question $question;
```
This is not used anywhere, right?

adevr replied:
You're right! I can remove the question attribute from the controller.

tiagof commented on domains/Discussions/Controllers/AnswersUpdateController.php:
```diff
@@ -0,0 +1,42 @@
+
+    public function __invoke(int $questionId, int $answerId, Request $request): Response
+    {
+        $answer = $this->answer->where("question_id", $questionId)->findOrFail($answerId);
```
Let's use single quote to make it consistent.
Also the `where(..)` is not needed since we're targeting a single answer with `findOrFail`

adevr replied:
This will be impacted by the answer to your question on the issue. #33

tiagof replied:
@adevr @josepostiga already commented on this below. So please keep the where and just use the single quote.

tiagof commented on domains/Discussions/Controllers/AnswersUpdateController.php:
```diff
@@ -0,0 +1,42 @@
+
+        $answer->update([
+            'content' => $request->input("content"),
+            'update_at' => Carbon::now()
```
Should be `updated_at` but this is really not needed since eloquent takes care of that automatically.

adevr replied:
Yes, I forgot that behaviour! This will be fixed

tiagof commented on domains/Discussions/Policies/AnswerPolicy.php:
```diff
@@ -0,0 +1,19 @@
+{
+    use HandlesAuthorization;
+
+    public function update(User $user, Answer $question): bool
```
Could you change the 2nd param name to `$answer` to avoid confusion :) ?

tiagof commented on domains/Discussions/Tests/Feature/AnswersUpdateTest.php:
```diff
@@ -0,0 +1,106 @@
+        parent::setUp();
+
+        $this->faker = Factory::create();
+        $this->user = UserFactory::new()->create();
+        $this->question = QuestionFactory::new(['author_id' => $this->user->id])->create();
+        $this->answer = AnswerFactory::new([
+            'author_id' => $this->user->id,
+            'question_id' => $this->question->id,
+        ])->create();
```
Suggestion:

maybe avoid calling all these Factories by calling the AnswerFactory first
`$this->answer = AnswerFactory::new()->create()`
and then do
`$this->user = $this->answer->author`
and
`$this->question = $this->answer->question`

adevr replied:
You're right!

tiagof commented on domains/Discussions/Tests/Feature/AnswersUpdateTest.php:
```diff
@@ -0,0 +1,106 @@
+        ];
+
+        $response = $this->actingAs($this->user)
+            ->call(
+                'PATCH',
```
Use `->patch(route(...), $payload)` to keep it consistent with the tests below.

tiagof reviewed (commented):
Hi @adevr thanks for this. Great job!
I've left a few comments below just to keep it consistent with the project.
More importantly I've left a question (https://github.com/laravel-portugal/api/issues/33#issuecomment-722527262) which the asnwer to can impact some of these comments.

adevr reviewed (commented):
Fixed

josepostiga commented on domains/Discussions/Models/Answer.php:
```diff
@@ -13,12 +13,15 @@ class Answer extends Model
             ->withTrashed();
     }
 
+
```
Extra space, here. Care to remove it?

josepostiga commented on domains/Discussions/Policies/AnswerPolicy.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function update(User $user, Answer $answer): bool
+    {
+        return $answer->author->is($user);
```
We could take advantage of the question we're loading in the controller, pass it to the policy and make sure that it's the one related to the answer we're trying to update, forbidding the operation otherwise.

josepostiga reviewed (changes requested):
Thanks for this @adevr. There are a few more things we need to work out, but we're getting close to making this PR ready for merging.

adevr commented:
@tiagof @josepostiga Let me know if it needs more improvement

josepostiga commented on domains/Discussions/Controllers/AnswersUpdateController.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function __invoke(Request $request, int $questionId, int $answerId): Response
+    {
+        $answer = $this->answer->OfQuestion($questionId)->findOrFail($answerId);
```
The scope name must be camelCased. Also, with that scope, there's no need to search for the question, specifically.

josepostiga commented on domains/Discussions/Policies/AnswerPolicy.php:
```diff
@@ -0,0 +1,20 @@
+
+    public function update(User $user, Answer $answer, Question $question): bool
+    {
+        return $answer->author->is($user) && $answer->question->is($question);
```
With the scope we're applying in the controller, there's no need to check that the answer is related to the question. The generated query already checks that, implicitly.

adevr replied:
I was pretty confused that but I think I fixed what you wanted! Please look at my last commit.
At some point in the past I remember reading to do this according to that, but I adjusted this point anyway.

josepostiga replied:
Yes, it was an oversight on my part as I didn't quite notice the scope was already validating that the given answer must be associated with the given question.

josepostiga reviewed (changes requested):
We're almost there! I just noted some things we can do to improve the solution but, after that, I believe we can merge this :+1: 

Good work!

josepostiga reviewed (changes requested):
It all looks ok. I think we need to only add a test case to validate that no one can update other people's answers.

adevr commented:
Well I added the test and I fixed one of my commit messages. Also I added a line inside the changelog separator @josepostiga

adevr commented:
> Well I added the test and I fixed one of my commit messages. Also I added a line inside the changelog separator @josepostiga

Let me know if it needs more work

josepostiga reviewed (approved):
Good work :+1:
